### ISSUE TITLE
Significant deploy time increase

### ISSUE THREAD

**Issue by JalisoCSP** [Shopify/bootsnap#178]
EDIT: This seemed to occur with `require 'bootsnap/setup'` rather than `Bootsnap.setup`, as when we noticed require was only `'bootsnap'` instead of `'bootsnap/setup'`, the problem persisted - changing to only require when ENV is development seemed to work for us...

I see some potentially related issues, but I'm hoping this will hope out somewhat.

We're seeing significant increases in deploy times while using the following code in config/boot.rb
```
Bootsnap.setup(
  cache_dir:            'tmp/cache',          # Path to your cache
  development_mode:     env == 'development', # Current working environment, e.g. RACK_ENV, RAILS_ENV, etc
  load_path_cache:      true,                 # Optimize the LOAD_PATH with a cache
  autoload_paths_cache: true,                 # Optimize ActiveSupport autoloads with cache
  disable_trace:        true,                 # (Alpha) Set `RubyVM::InstructionSequence.compile_option = { trace_instruction: false }`
  compile_cache_iseq:   false,                # Compile Ruby code into ISeq cache, breaks coverage reporting.
  compile_cache_yaml:   true                  # Compile YAML into a cache
)
```
It seemed like our issue with going progressively worse and worse adding almost 15 minutes to each deploy, every time we deployed.

We noticed the issue on assets:precompile for a while, before getting frustrated and removing the precompile step in our deploy and seeing the same issue occur on db:migrate. Removing both of these steps reduced our deploy time to roughly 1 minute.

This lead us to notice that both of these commands use `bundle exec rake`, but we got no further in the cause or how to fix this without removing Bootsnap.setup - the only reason this was added, was to fix code coverage, so we felt this was fine.

This happened to us on Rails 5.1.6 AND Rails 5.2.0, both with Ruby 2.5.1 and deploying with Capistrano.

Hope this helps, but let me know if you need some more information.

**Comment by npezza93** [comment]
I have experienced this as well on Rails 5.0.7 and ruby 2.4.1. Our deploys went from 1 minute to 20.
Changing 
```ruby
require 'bundler/setup'
require "bootsnap/setup"
```
to 
```ruby
require 'bundler/setup'
if Rails.env.development? 
  require "bootsnap/setup"
end
``` 
resolved the issue.

**Comment by JalisoCSP** [comment]
@npezza93 yep, I said about checking for the dev environment in my edit. Doesn't seem to be required and there is no mention of this in the docs, so I wonder if there is another reason. Happy to close this though.

**Comment by divoxx** [comment]
Was there a specific reason why this was closed? The README states that bootsnap is safe to be used in production and that Shopify does, therefore this should probably be discussed and potentially addressed by the developers since disabling it in production shouldn't be a requirement.

**Comment by mjgreer** [comment]
we also had the same issue in prod and found it was due to a huge cache being built by bootsnap since it saw every deploy directory as a new file to cache. @JalisoCSP, is there a chance this could be reopened?

**Comment by burke** [comment]
If Capistrano is involved, I bet this has to do with a continually-growing cache. I'll document better in the README that we leave it up to the user to clean the cache occasionally.

**Comment by mjgreer** [comment]
@burke it seemed that it was seeing every new release directory as a 'new' file, regardless of changes  in this case, wouldn't that nullify the benefit of bootsnap?

**Comment by burke** [comment]
Yeah it would, I guess you're checking out entirely new copies of the files for each revision. We don't use capistrano anymore so I haven't bothered to optimize for that use-case. 

This is the cache key definition: https://github.com/Shopify/bootsnap/blob/master/ext/bootsnap/bootsnap.c#L32-L55

I guess the `mtime` is probably what's invalidating it. One way to force it to sync would be to set the mtime of all of the ruby files to 0 (or any other constant).

**Comment by rahim** [comment]
> we also had the same issue in prod and found it was due to a huge cache being built by bootsnap since it saw every deploy directory as a new file to cache. @JalisoCSP, is there a chance this could be reopened?

I've also observed `assets:precompile` runtime growing to excessive levels during a capistrano deploy - `strace` showed the process was repeatedly stalling on the `stat` syscall for `./tmp/cache` with the directory containing large numbers of files, in this particular case bootsnap had generated a few hundred thousand in the space of a few weeks. Manually purging the cache directory brought runtime back down to sensible levels.

Given bootsnap is included by default in production for new Rails applications, it feels dangerous to put the burden of periodically clearing the cache on developers (however sensible this is), it's unlikely many will even realise the cause of the slowdown is cache growth let alone read the bootstrap documentation.

Would it be reasonable to have bootsnap prune its own cache during boot?

**Comment by maxigs** [comment]
Been looking into deploy time increases on a rails app with capistrano for months now without luck.

Quick summary of what fixed the issue now:

In the deploy config for capistrano i removed `tmp/cache` from the `linked_dirs`.
If neede specific cache sub-folders can be re-added, but bootsnap content has to stay ouf of it.

Now each deployment resets the cache, but not clearing out a shared one while old processes are still running.
Capistrano then takes care of removing the cache with the deploy artifact automatically.

Deploys are back to normal speed now, hopefully this time it will remain like this.

For who is interested in the failed tries before i found this issue:

1. Moved asset precompile to CI and rsync them to the server (not a bad idea in general) -> helped a little while, but no luck
2. Upgraded Ruby, hoping it would help speed up the growing app -> no luck
3. Upgraded the server to a way oversized one, hoping it was an issue with cpu-credits or some IO constraint -> helped a for a little while, but no luck
4. Fiddled around with migrations, as the migrate was the last blocking process on deploy -> no luck
5. Found this bootsnap issue (by accident, looking over the gemfile, trying to offload some legacy dependencies) -> 🍾 🎆
